doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py

```diff
@@ -611,23 +624,20 @@ def main():
             use_gpu=True,
             resources_per_worker={"GPU": 1},
         ),
-        datasets={
-            "train": train_ds,
-            "valid": valid_ds,
-        },
-        dataset_config=ray.train.DataConfig(
-            datasets_to_split=["train", "valid"],
-        ),
+        datasets={"train": train_ds, "valid": valid_ds},
+        dataset_config=ray.train.DataConfig(datasets_to_split=["train", "valid"]),
     )
 
-    results = trainer.fit()
-    best_checkpoint = results.best_checkpoints[0]
+    result: train.Result = trainer.fit()
+    # `best_checkpoints` are sorted in increasing score order.
+    # (Ex: in this case, negative perplexity, since we set `checkpoint_score_order=min`)
+    best_checkpoint, best_checkpoint_metrics = result.best_checkpoints[-1]
 
-    print("Results are stored in:")
-    print(results.path)
-    print("Best checkpoint is stored in:")
-    print(best_checkpoint[0].uri)
-    print(f"With perplexity: {best_checkpoint[1]['perplexity']}")
+    print("Results are stored at:")
+    print(result.path)
+    print("Best checkpoint is stored at:")
+    print(best_checkpoint)
+    print(f"With perplexity: {best_checkpoint_metrics['perplexity']}")
```

Comment:
I think the worst checkpoint was being reported previously 😅

Reply:
oh interesting :)